From: Serge Hallyn <serge.hallyn@ubuntu.com>
Date: Wed, 13 Nov 2013 13:44:05 -0600
Subject: accomodate new qemu migration status 'setup'

Treat it as active, but don't try to query for status info
yet.

Signed-off-by: Serge Hallyn <serge.hallyn@ubuntu.com>
---
 src/qemu/qemu_migration.c    | 48 ++++++++++++++++++++++++++++----------------
 src/qemu/qemu_monitor.c      |  8 +++++---
 src/qemu/qemu_monitor.h      |  3 ++-
 src/qemu/qemu_monitor_json.c | 13 +++++++++---
 src/qemu/qemu_monitor_json.h |  3 ++-
 src/qemu/qemu_monitor_text.c |  9 ++++++++-
 src/qemu/qemu_monitor_text.h |  3 ++-
 tests/qemumonitorjsontest.c  |  4 ++--
 8 files changed, 62 insertions(+), 29 deletions(-)

diff --git a/src/qemu/qemu_migration.c b/src/qemu/qemu_migration.c
index 54c6fec..e917ba7 100644
--- a/src/qemu/qemu_migration.c
+++ b/src/qemu/qemu_migration.c
@@ -1661,7 +1661,7 @@ qemuMigrationUpdateJobStatus(virQEMUDriverPtr driver,
                              enum qemuDomainAsyncJob asyncJob)
 {
     qemuDomainObjPrivatePtr priv = vm->privateData;
-    int ret;
+    int ret, setting_up;
     qemuMonitorMigrationStatus status;
 
     memset(&status, 0, sizeof(status));
@@ -1671,7 +1671,7 @@ qemuMigrationUpdateJobStatus(virQEMUDriverPtr driver,
         /* Guest already exited; nothing further to update.  */
         return -1;
     }
-    ret = qemuMonitorGetMigrationStatus(priv->mon, &status);
+    ret = qemuMonitorGetMigrationStatus(priv->mon, &status, &setting_up);
 
     qemuDomainObjExitMonitor(driver, vm);
 
@@ -1696,21 +1696,35 @@ qemuMigrationUpdateJobStatus(virQEMUDriverPtr driver,
         break;
 
     case QEMU_MONITOR_MIGRATION_STATUS_ACTIVE:
-        priv->job.info.fileTotal = priv->job.status.disk_total;
-        priv->job.info.fileRemaining = priv->job.status.disk_remaining;
-        priv->job.info.fileProcessed = priv->job.status.disk_transferred;
-
-        priv->job.info.memTotal = priv->job.status.ram_total;
-        priv->job.info.memRemaining = priv->job.status.ram_remaining;
-        priv->job.info.memProcessed = priv->job.status.ram_transferred;
-
-        priv->job.info.dataTotal =
-            priv->job.status.ram_total + priv->job.status.disk_total;
-        priv->job.info.dataRemaining =
-            priv->job.status.ram_remaining + priv->job.status.disk_remaining;
-        priv->job.info.dataProcessed =
-            priv->job.status.ram_transferred +
-            priv->job.status.disk_transferred;
+        if (setting_up) {
+            priv->job.info.fileTotal = -1;
+            priv->job.info.fileRemaining = -1;
+            priv->job.info.fileProcessed = 0;
+
+            priv->job.info.memTotal = -1;
+            priv->job.info.memRemaining = -1;
+            priv->job.info.memProcessed = 0;
+
+            priv->job.info.dataTotal = -1;
+            priv->job.info.dataRemaining = -1;
+            priv->job.info.dataProcessed = 0;
+        } else {
+            priv->job.info.fileTotal = priv->job.status.disk_total;
+            priv->job.info.fileRemaining = priv->job.status.disk_remaining;
+            priv->job.info.fileProcessed = priv->job.status.disk_transferred;
+
+            priv->job.info.memTotal = priv->job.status.ram_total;
+            priv->job.info.memRemaining = priv->job.status.ram_remaining;
+            priv->job.info.memProcessed = priv->job.status.ram_transferred;
+
+            priv->job.info.dataTotal =
+                priv->job.status.ram_total + priv->job.status.disk_total;
+            priv->job.info.dataRemaining =
+                priv->job.status.ram_remaining + priv->job.status.disk_remaining;
+            priv->job.info.dataProcessed =
+                priv->job.status.ram_transferred +
+                priv->job.status.disk_transferred;
+        }
 
         ret = 0;
         break;
diff --git a/src/qemu/qemu_monitor.c b/src/qemu/qemu_monitor.c
index a2769db..708d899 100644
--- a/src/qemu/qemu_monitor.c
+++ b/src/qemu/qemu_monitor.c
@@ -2095,7 +2095,8 @@ qemuMonitorSetMigrationCacheSize(qemuMonitorPtr mon,
 
 
 int qemuMonitorGetMigrationStatus(qemuMonitorPtr mon,
-                                  qemuMonitorMigrationStatusPtr status)
+                                  qemuMonitorMigrationStatusPtr status,
+                                  int *setting_up)
 {
     int ret;
     VIR_DEBUG("mon=%p", mon);
@@ -2106,10 +2107,11 @@ int qemuMonitorGetMigrationStatus(qemuMonitorPtr mon,
         return -1;
     }
 
+    *setting_up = 0;
     if (mon->json)
-        ret = qemuMonitorJSONGetMigrationStatus(mon, status);
+        ret = qemuMonitorJSONGetMigrationStatus(mon, status, setting_up);
     else
-        ret = qemuMonitorTextGetMigrationStatus(mon, status);
+        ret = qemuMonitorTextGetMigrationStatus(mon, status, setting_up);
     return ret;
 }
 
diff --git a/src/qemu/qemu_monitor.h b/src/qemu/qemu_monitor.h
index eabf000..e6c93e3 100644
--- a/src/qemu/qemu_monitor.h
+++ b/src/qemu/qemu_monitor.h
@@ -434,7 +434,8 @@ struct _qemuMonitorMigrationStatus {
 };
 
 int qemuMonitorGetMigrationStatus(qemuMonitorPtr mon,
-                                  qemuMonitorMigrationStatusPtr status);
+                                  qemuMonitorMigrationStatusPtr status,
+                                  int *setting_up);
 int qemuMonitorGetSpiceMigrationStatus(qemuMonitorPtr mon,
                                        bool *spice_migrated);
 
diff --git a/src/qemu/qemu_monitor_json.c b/src/qemu/qemu_monitor_json.c
index 5e825ac..4aeceef 100644
--- a/src/qemu/qemu_monitor_json.c
+++ b/src/qemu/qemu_monitor_json.c
@@ -2353,7 +2353,8 @@ qemuMonitorJSONSetMigrationCacheSize(qemuMonitorPtr mon,
 
 static int
 qemuMonitorJSONGetMigrationStatusReply(virJSONValuePtr reply,
-                                       qemuMonitorMigrationStatusPtr status)
+                                       qemuMonitorMigrationStatusPtr status,
+                                       int *setting_up)
 {
     virJSONValuePtr ret;
     const char *statusstr;
@@ -2371,6 +2372,11 @@ qemuMonitorJSONGetMigrationStatusReply(virJSONValuePtr reply,
         return -1;
     }
 
+    if (strncmp(statusstr, "setup", 5) == 0) {
+        status->status = QEMU_MONITOR_MIGRATION_STATUS_ACTIVE;
+        *setting_up = 1;
+        return 0;
+    }
     status->status = qemuMonitorMigrationStatusTypeFromString(statusstr);
     if (status->status < 0) {
         virReportError(VIR_ERR_INTERNAL_ERROR,
@@ -2511,7 +2517,8 @@ qemuMonitorJSONGetMigrationStatusReply(virJSONValuePtr reply,
 
 
 int qemuMonitorJSONGetMigrationStatus(qemuMonitorPtr mon,
-                                      qemuMonitorMigrationStatusPtr status)
+                                      qemuMonitorMigrationStatusPtr status,
+                                      int *setting_up)
 {
     int ret;
     virJSONValuePtr cmd = qemuMonitorJSONMakeCommand("query-migrate",
@@ -2529,7 +2536,7 @@ int qemuMonitorJSONGetMigrationStatus(qemuMonitorPtr mon,
         ret = qemuMonitorJSONCheckError(cmd, reply);
 
     if (ret == 0 &&
-        qemuMonitorJSONGetMigrationStatusReply(reply, status) < 0)
+        qemuMonitorJSONGetMigrationStatusReply(reply, status, &setting_up) < 0)
         ret = -1;
 
     if (ret < 0)
diff --git a/src/qemu/qemu_monitor_json.h b/src/qemu/qemu_monitor_json.h
index a93c51e..6ed464d 100644
--- a/src/qemu/qemu_monitor_json.h
+++ b/src/qemu/qemu_monitor_json.h
@@ -131,7 +131,8 @@ int qemuMonitorJSONSetMigrationCacheSize(qemuMonitorPtr mon,
                                          unsigned long long cacheSize);
 
 int qemuMonitorJSONGetMigrationStatus(qemuMonitorPtr mon,
-                                      qemuMonitorMigrationStatusPtr status);
+                                      qemuMonitorMigrationStatusPtr status,
+                                      int *setting_up);
 
 int qemuMonitorJSONGetMigrationCapability(qemuMonitorPtr mon,
                                           qemuMonitorMigrationCaps capability);
diff --git a/src/qemu/qemu_monitor_text.c b/src/qemu/qemu_monitor_text.c
index 99f7641..42edf4c 100644
--- a/src/qemu/qemu_monitor_text.c
+++ b/src/qemu/qemu_monitor_text.c
@@ -1419,7 +1419,8 @@ cleanup:
 #define MIGRATION_DISK_TOTAL_PREFIX "total disk: "
 
 int qemuMonitorTextGetMigrationStatus(qemuMonitorPtr mon,
-                                      qemuMonitorMigrationStatusPtr status)
+                                      qemuMonitorMigrationStatusPtr status,
+                                      int *setting_up)
 {
     char *reply;
     char *tmp;
@@ -1441,6 +1442,12 @@ int qemuMonitorTextGetMigrationStatus(qemuMonitorPtr mon,
         }
         *end = '\0';
 
+        if (strncmp(tmp, "setup", 5) == 0) {
+            status->status = QEMU_MONITOR_MIGRATION_STATUS_ACTIVE;
+            *setting_up = 1;
+            goto done;
+        }
+
         status->status = qemuMonitorMigrationStatusTypeFromString(tmp);
         if (status->status < 0) {
             virReportError(VIR_ERR_INTERNAL_ERROR,
diff --git a/src/qemu/qemu_monitor_text.h b/src/qemu/qemu_monitor_text.h
index 5218a8b..8441c08 100644
--- a/src/qemu/qemu_monitor_text.h
+++ b/src/qemu/qemu_monitor_text.h
@@ -117,7 +117,8 @@ int qemuMonitorTextSetMigrationDowntime(qemuMonitorPtr mon,
                                         unsigned long long downtime);
 
 int qemuMonitorTextGetMigrationStatus(qemuMonitorPtr mon,
-                                      qemuMonitorMigrationStatusPtr status);
+                                      qemuMonitorMigrationStatusPtr status,
+                                      int *setting_up);
 
 int qemuMonitorTextMigrate(qemuMonitorPtr mon,
                            unsigned int flags,
diff --git a/tests/qemumonitorjsontest.c b/tests/qemumonitorjsontest.c
index d7da5a8..c1fb198 100644
--- a/tests/qemumonitorjsontest.c
+++ b/tests/qemumonitorjsontest.c
@@ -1661,7 +1661,7 @@ testQemuMonitorJSONqemuMonitorJSONGetMigrationStatus(const void *data)
 {
     virDomainXMLOptionPtr xmlopt = (virDomainXMLOptionPtr)data;
     qemuMonitorTestPtr test = qemuMonitorTestNewSimple(true, xmlopt);
-    int ret = -1;
+    int ret = -1, setting_up;
     qemuMonitorMigrationStatus status, expectedStatus;
 
     if (!test)
@@ -1690,7 +1690,7 @@ testQemuMonitorJSONqemuMonitorJSONGetMigrationStatus(const void *data)
                                "}") < 0)
         goto cleanup;
 
-    if (qemuMonitorJSONGetMigrationStatus(qemuMonitorTestGetMonitor(test), &status) < 0)
+    if (qemuMonitorJSONGetMigrationStatus(qemuMonitorTestGetMonitor(test), &status, &setting_up) < 0)
         goto cleanup;
 
     if (memcmp(&status, &expectedStatus, sizeof(status)) != 0) {
